src/util/authtok-utils.c

```diff
@@ -72,3 +72,90 @@ errno_t sss_auth_pack_2fa_blob(const char *fa1, size_t fa1_len,
 
     return 0;
 }
+
+errno_t sss_auth_pack_sc_blob(const char *pin, size_t pin_len,
+                              const char *token_name, size_t token_name_len,
+                              const char *module_name, size_t module_name_len,
+                              const char *key_id, size_t key_id_len,
+                              uint8_t *buf, size_t buf_len,
+                              size_t *_sc_blob_len)
+{
+    size_t c;
+    uint32_t tmp_uint32_t;
+
+    if (pin_len > UINT32_MAX || token_name_len > UINT32_MAX
+            || module_name_len > UINT32_MAX
+            || (pin_len != 0 && pin == NULL)
+            || (token_name_len != 0 && token_name == NULL)
+            || (module_name_len != 0 && module_name == NULL)
+            || (key_id_len != 0 && key_id == NULL)) {
+        return EINVAL;
+    }
+
+    /* A missing pin is ok in the case of a reader with a keyboard */
+    if (pin == NULL) {
+        pin = "";
```

Comment:
it would be nice to set the pin_len to strlen(pin) here to avoid issues if someone passed in NULL pin but non-zero pin_len. Same with other input strings.